pkg/dump/pipeline/pipeline.go

```diff
@@ -0,0 +1,153 @@
+	operationName string
+	entity        string
+	streamFunc    StreamFunc
+}
+
+// dumpIngestorSequence returns the pipeline sequence for dumping k8s object (can be multi-threaded depending on the writer used)
+func dumpIngestorSequence(collector collector.CollectorClient, writer writer.DumperWriter) []DumpIngestorPipeline {
+	return []DumpIngestorPipeline{
+		{
+			operationName: span.DumperNodes,
+			entity:        tag.EntityNodes,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamNodes(ctx, NewNodeIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperPods,
+			entity:        tag.EntityPods,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamPods(ctx, NewPodIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperRoles,
+			entity:        tag.EntityRoles,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamRoles(ctx, NewRoleIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperClusterRoles,
+			entity:        tag.EntityClusterRoles,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamClusterRoles(ctx, NewClusterRoleIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperRoleBindings,
+			entity:        tag.EntityRolebindings,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamRoleBindings(ctx, NewRoleBindingIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperClusterRoleBindings,
+			entity:        tag.EntityClusterRolebindings,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamClusterRoleBindings(ctx, NewClusterRoleBindingIngestor(ctx, writer))
+			},
+		},
+		{
+			operationName: span.DumperEndpoints,
+			entity:        tag.EntityEndpoints,
+			streamFunc: func(ctx context.Context) error {
+				return collector.StreamEndpoints(ctx, NewEndpointIngestor(ctx, writer))
+			},
+		},
+	}
+}
+
+// PipelineDumpIngestor is a parallelized pipeline based ingestor implementation.
+type PipelineDumpIngestor struct {
+	sequence     []DumpIngestorPipeline
+	wp           worker.WorkerPool
+	WorkerNumber int
+}
+
+func NewPipelineDumpIngestor(ctx context.Context, collector collector.CollectorClient, writer writer.DumperWriter) (context.Context, *PipelineDumpIngestor, error) {
+	sequence := dumpIngestorSequence(collector, writer)
+
+	// Getting the number of workers from the writer to setup multi-threading if possible
+	workerNumber := writer.WorkerNumber()
+	// Set the number of workers to the number of differents entities (roles, pods, ...)
+	if workerNumber == 0 {
+		workerNumber = len(sequence)
+	}
+
+	if workerNumber > 1 {
+		log.I.Infof("Multi-threading enabled: %d workers", workerNumber)
+	}
+
+	// Setting up the worker pool with multi-threading if possible
+	// enable for raw file writer
+	// disable for targz writer (not thread safe)
+	wp, err := worker.PoolFactory(workerNumber, 1)
```

Comment:
is 1 intended for worker > 1?
Does this actually run in multiple threads if you put workerNumber to like 5 but with a capacity of 1?
(Also, I would suggest having that "1" put into a variable, because that's not easily understandable, something like poolCapacity or something maybe?)

Reply:
yes, this is the capacity of the queue. In our case we do not need any buffer. I will add variable to make it clearier.

Reply:
I'm still unclear on _why_ we don't want any more buffer, but LGTM!
(I guess we will see on the trace how much concurrency we are really getting. In any case I would assume it's not a big deal :D)

Reply:
it was to avoid any concurrency issue. I guess I can try to test it in another PR to see if perf improve and does not introduce new bugs.